CMakeLists.txt

```diff
@@ -1447,7 +1447,7 @@ if(CONFIG_BUILD_OUTPUT_BIN AND CONFIG_BUILD_OUTPUT_UF2)
     COMMAND ${PYTHON_EXECUTABLE} ${ZEPHYR_BASE}/scripts/uf2conv.py
             -c
             -f ${CONFIG_BUILD_OUTPUT_UF2_FAMILY_ID}
-            -b ${CONFIG_FLASH_LOAD_OFFSET}
+            -b ${CONFIG_BUILD_OUTPUT_UF2_OFFSET}
```

Comment:
According to the commit log, all I should expect this patch to do is "Generate a UF2 file when building for the rpi_pico."

But it's doing more than that; in fact it's changing the offset passed to uf2conf.py right here. That change is not justified in the commit log, though, so I don't know how to review it.

Is the old CONFIG option a bug? Then it should probably be fixed in its own commit with a good justification; mixing bug fixes in commits that introduce new functionality makes things harder to review and understand, especially without justifications in the commit log.

Or is something else going on? If the old value is not a bug, and it's something else, then I'm not sure what it is based on what's in the patch itself. Please remember that the commit history is the main thing we have to go back to when trying to figure out why things happened, so it's worth spending time on structuring them and writing good log comments for them.

Reply:
I had created three unique commits for this work in my branch, they may have gotten squashed.

The offset needed by the bootloader is *not* the same as the code offset.

The bootloader expects the actual physical flash location, where as the DT code is just the offset *from the flash location start*

Reply:
So this is a bug, and CONFIG_BUILD_OUTPUT_UF2_OFFSET is the absolute address this tool always needs on all platforms?

Reply:
@mbolivar-nordic As I think about it.... I think if we consistently used:

* Flash register address + FLASH_LOAD_OFFSET when `CONFIG_USE_DT_CODE_PARTITION=y`, otherwise
* Just use flash register address

We would at least be right for Adafruit nRF52 bootloader, samd21 bootloader and RP2040. 

So, That might actually be the right way to handle this, the question would be implementation:

The `uf2conv.py` script is an external script, expects a single offset value passed on the command line. Can we use CMake's `math` here and be ok? I'm not a CMake expert.

Reply:
@petejohanson Note that for rp2040, it's flash_base if requires_bootloader, else flash_base + flash_load_offset

Reply:
Wouldn't we just select `USE_DT_CODE_OFFSET` for that bootloader scenario? Seems like that would keep it consistent and simpler.

Reply:
@petejohanson Do you mean disable `USE_DT_CODE_PARTITION` if a bootloader is present? That could work

Reply:
> Can we use CMake's math here and be ok?

I don't see any reason why not; if you're confident that works across platforms it sounds reasonable to me.

Reply:
Just please make sure to describe the change thoroughly in the commit log on next update, of course, so this information doesn't get lost in the PR thread :)

Reply:
@mbolivar-nordic I pushed the better implementation of this into #42210 that hasn't yet been incorprated here.

With that in place, the offset wouldn't be hardcoded, and this RP2 stuff would work by:

1) Add `CONFIG_USE_DT_CODE_PARTITION=y` to the board's `<board>_defconfig` file.
2) Adding: `config BUILD_OUTPUT_UF2_IGNORE_FLASH_OFFSET\ndefault y` to the RP2 SoC `Kconfig.defconfig` file.

The toplevel `CMakeLists.txt` logic would then pull the configured flash base address, ignore the 0x100 flash offset, and build the UF2 with a base address of 0x100000 which is what's needed for the combined second stage bootloader + Zephyr application.

Reply:
> > Can we use CMake's math here and be ok?
> 
> I don't see any reason why not; if you're confident that works across platforms it sounds reasonable to me.

I actually found it used already elsewhere in a linker.cmake file, and basically copied it directly for my change in #42210, so it seems safe to me.

Reply:
Sure, using `math` is fine -- I meant if you're confident about the results of the calculations being correct across target platforms, that seems fine to me.

Reply:
yep, using `math` is fine.